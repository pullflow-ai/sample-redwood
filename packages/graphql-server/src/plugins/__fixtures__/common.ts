import { makeExecutableSchema } from '@graphql-tools/schema'

export const testSchema = makeExecutableSchema({
  typeDefs: /* GraphQL */ `
    type Query {
      me: User!
    }

    type Query {
      forbiddenUser: User!
      getUser(id: Int!): User!
    }

    type User {
      id: ID!
      name: String!
    }
  `,
  resolvers: {
    Query: {
      me: () => {
        return { _id: 1, firstName: 'Ba', lastName: 'Zinga' }
      },
      forbiddenUser: () => {
        throw Error('You are forbidden')
      },
      getUser: (id) => {
        return { id, firstName: 'Ba', lastName: 'Zinga' }
      },
    },
    User: {
      id: (u) => u._id,
      name: (u) => `${u.firstName} ${u.lastName}`,
    },
  },
})

export const testLiveSchema = makeExecutableSchema({
  typeDefs: /* GraphQL */ `
    """
    Instruction for establishing a live connection that is updated once the underlying data changes.
    """
    directive @live(
      """
      Whether the query should be live or not.
      """
      if: Boolean = true

      """
      Propose a desired throttle interval ot the server in order to receive updates to at most once per "throttle" milliseconds. The server must not accept this value.
      """
      throttle: Int
    ) on QUERY

    type Query {
      me: User!
    }

    type Query {
      forbiddenUser: User!
      getUser(id: Int!): User!
    }

    type User {
      id: ID!
      name: String!
    }
  `,
  resolvers: {
    Query: {
      me: () => {
        return { _id: 1, firstName: 'Ba', lastName: 'Zinga' }
      },
      forbiddenUser: () => {
        throw Error('You are forbidden')
      },
      getUser: (id) => {
        return { id, firstName: 'Ba', lastName: 'Zinga' }
      },
    },
    User: {
      id: (u) => u._id,
      name: (u) => `${u.firstName} ${u.lastName}`,
    },
  },
})
export const testQuery = /* GraphQL */ `
  query meQuery {
    me {
      id
      name
    }
  }
`

export const testFilteredQuery = /* GraphQL */ `
  query FilteredQuery {
    me {
      id
      name
    }
  }
`

export const testErrorQuery = /* GraphQL */ `
  query forbiddenUserQuery {
    forbiddenUser {
      id
      name
    }
  }
`

export const testLiveQuery = /* GraphQL */ `
  query meQuery @live {
    me {
      id
      name
    }
  }
`

export const testParseErrorQuery = /* GraphQL */ `
  query ParseErrorQuery {
    me {
      id
      name
      unknown_field
    }
  }
`

export const testValidationErrorQuery = /* GraphQL */ `
  query ValidationErrorQuery(id: Int!) {
    getUser(id: 'one') {
      id
      name
    }
  }
`
