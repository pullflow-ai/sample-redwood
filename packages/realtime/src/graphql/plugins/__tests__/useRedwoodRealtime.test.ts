import {
  createTestkit,
  createSpiedPlugin,
  assertStreamExecutionValue,
} from '@envelop/testing'
import { describe, it, expect } from 'vitest'

import { testQuery, testLiveQuery, testSchema } from '../__fixtures__/common'
import {
  useRedwoodRealtime,
  InMemoryLiveQueryStore,
} from '../useRedwoodRealtime'

describe('useRedwoodRealtime', () => {
  it('should support a @live query directive', async () => {
    const testkit = createTestkit(
      [useRedwoodRealtime({ liveQueries: { store: 'in-memory' } })],
      testSchema,
    )

    const result = await testkit.execute(testLiveQuery, {}, {})

    assertStreamExecutionValue(result)
    const current = await result.next()
    expect(current.value).toMatchInlineSnapshot(`
      {
        "data": {
          "me": {
            "id": "1",
            "name": "Ba Zinga",
          },
        },
        "isLive": true,
      }
    `)
  })

  it('should update schema with live directive', () => {
    const spiedPlugin = createSpiedPlugin()

    // the original schema should not have the live directive before the useRedwoodRealtime plugin is applied
    expect(testSchema.getDirective('live')).toBeUndefined()

    createTestkit(
      [
        useRedwoodRealtime({ liveQueries: { store: 'in-memory' } }),
        spiedPlugin.plugin,
      ],
      testSchema,
    )

    // the replaced schema should have the live directive afterwards
    const replacedSchema =
      spiedPlugin.spies.onSchemaChange.mock.calls[0][0].schema

    const liveDirectiveOnSchema = replacedSchema.getDirective('live')

    expect(liveDirectiveOnSchema.name).toEqual('live')
    expect(replacedSchema.getDirective('live')).toMatchSnapshot()
  })

  it('with live directives, it should extend the graphQL context with a store', async () => {
    const spiedPlugin = createSpiedPlugin()

    const testkit = createTestkit(
      [
        useRedwoodRealtime({ liveQueries: { store: 'in-memory' } }),
        spiedPlugin.plugin,
      ],
      testSchema,
    )

    await testkit.execute(testQuery)

    expect(spiedPlugin.spies.beforeContextBuilding).toHaveBeenCalledTimes(1)
    expect(spiedPlugin.spies.beforeContextBuilding).toHaveBeenCalledWith({
      context: expect.any(Object),
      extendContext: expect.any(Function),
      breakContextBuilding: expect.any(Function),
    })

    expect(spiedPlugin.spies.afterContextBuilding).toHaveBeenCalledTimes(1)
    expect(spiedPlugin.spies.afterContextBuilding).toHaveBeenCalledWith({
      context: expect.objectContaining({
        liveQueryStore: expect.any(InMemoryLiveQueryStore),
      }),
      extendContext: expect.any(Function),
    })
  })

  it('with subscriptions, it should extend the GraphQL context with pubSub transport', async () => {
    const spiedPlugin = createSpiedPlugin()

    const testkit = createTestkit(
      [
        useRedwoodRealtime({
          subscriptions: { store: 'in-memory', subscriptions: [] },
        }),
        spiedPlugin.plugin,
      ],
      testSchema,
    )

    await testkit.execute(testQuery)

    expect(spiedPlugin.spies.beforeContextBuilding).toHaveBeenCalledTimes(1)
    expect(spiedPlugin.spies.beforeContextBuilding).toHaveBeenCalledWith({
      context: expect.any(Object),
      extendContext: expect.any(Function),
      breakContextBuilding: expect.any(Function),
    })

    expect(spiedPlugin.spies.afterContextBuilding).toHaveBeenCalledTimes(1)
    expect(spiedPlugin.spies.afterContextBuilding).toHaveBeenCalledWith({
      context: expect.objectContaining({
        pubSub: expect.objectContaining({
          publish: expect.any(Function),
          subscribe: expect.any(Function),
        }),
      }),
      extendContext: expect.any(Function),
    })
  })
})
