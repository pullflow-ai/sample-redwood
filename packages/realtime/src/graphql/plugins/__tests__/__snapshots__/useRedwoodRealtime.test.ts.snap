// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useRedwoodRealtime > should update schema with live directive 1`] = `"@live"`;
